Draft questions UI

jlevine18 opened the pull request:
UI for draft questions page. Changes AI generate + Jobs page.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.sql:
```diff
@@ -0,0 +1,23 @@
+  LEFT JOIN draft_question_metadata AS dqm ON dqm.question_id = q.id
+  LEFT JOIN users As u ON u.user_id = dqm.created_by
+WHERE
+  q.course_id = $course_id
```
any draft question will show (even if it is manually inserted); other columns will only show if they're recognized. we can change display format if needed.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.ts:
```diff
@@ -0,0 +1,78 @@
+        QuestionSchema,
+      );
+
+      for (const question of questions) {
```
we could also make a batch-delete editor and and remove all of the questions at once if that is preferred. would simply require modifying the trpc and making a new editor.

nwalters512 replied:
I think a batch-delete would be much better here. A full delete can be a really heavy operation for larger courses. I'd be entirely happy to save the whole implementation of deletes for another PR.

nwalters512 replied:
Deletion was indeed removed from this PR in f55db8a. We can add it back in the future.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -81,8 +94,43 @@ router.get(
   '/',
   asyncHandler(async (req, res) => {
     assertCanCreateQuestion(res.locals);
+    if (req.query?.qid) {
```
this might be a slightly weird decision but the questionEditor page and generateNewQuestion page are the same page, where the questionEditor page is simply when a valid qid is expressed in query-parameters. That way, if a user requests a question edit page with an invalid qid, we simply fall back to the new question generation page. we could just switch to making these separate pages and simply having a redirect elsewhere.

nwalters512 replied:
I think separate pages with a redirect would be much better. This would also allow us to embed the question ID / QID / whatever in the URL path (`.../:id`) rather than sticking it in the query string, which is unconventional for PL.

nwalters512 replied:
We're now using separate pages.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -81,8 +94,43 @@ router.get(
   asyncHandler(async (req, res) => {
     assertCanCreateQuestion(res.locals);
+    if (req.query?.qid) {
+      const qidFull = `__drafts__/${req.query?.qid}`;
```
we assume that all drafts questions start with `__drafts__` (by definition, they do!) so we keep that portion out of the query id and append it on.

nwalters512 replied:
Rather than all this messing around with QIDs, why not just use the question ID?

nwalters512 replied:
This change was ultimately made.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -81,8 +94,43 @@ router.get(
+          { qid: qidFull, course_id: res.locals.course.id },
+          QuestionSchema,
+        );
+        const variant_id = req.query.variant_id ? IdSchema.parse(req.query.variant_id) : null;
```
unfortunately, the renderer by default gets from '/' (while maintaining query parameters), so we have to do the parsing hacks here

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -81,8 +94,43 @@ router.get(
+        await logPageView('instructorQuestionPreview', req, res);
+        setRendererHeader(res);
+      }
+      const queryUrl = req.originalUrl.split('?')[1]; //only the info after the query seperator
```
we keep this parsed out so that we can do htmx requests with query parameters.

nwalters512 replied:
I think the need for this will go away if we stop using query params to store the QID, and instead make that part of the URL path itself.

nwalters512 replied:
This was indeed resolved by using the path to specify the question ID.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -113,14 +161,10 @@ router.post(
-          ),
-        );
+        res.set({
+          'HX-Redirect': `${res.locals.urlPrefix}/ai_generate_question?qid=${result.questionQid.substring(11)}`,
```
if we generate a valid question, redirect to the editor page for it.

jlevine18 commented on apps/prairielearn/src/api/trpc/routers/course-files/deleteQuestion.ts:
```diff
@@ -0,0 +1,70 @@
+
+    const question = await selectQuestionById(opts.input.question_id);
+
+    const editor = new QuestionDeleteEditor({
```
leeching off of single deletion editor for now even though we only use this rpc for batch deletions- see discussion below and lmk if you want me to change it.

nwalters512 replied:
I ended up removing bulk deletion for now in f55db8a. We can add it (and a broper batch-deletion endpoint) in another PR.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -158,20 +205,17 @@ router.post(
-          ),
-        );
+        res.set({
+          'HX-Redirect': `${res.locals.urlPrefix}/ai_generate_question?qid=${req.body.unsafe_qid}`,
```
redirect back to self- perhaps there's a better way to trigger page refresh from htmx? we need this to be an htmx request if we want to continue leeching off of loading-states.

nwalters512 replied:
This now redirects to another page.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -181,29 +225,59 @@ router.post(
 );
 
+router.get(
+  '/variant/:variant_id(\\d+)/submission/:submission_id(\\d+)',
```
this partial renders submission results (from backend grading).

nwalters512 replied:
Turns out this isn't necessary, it'll use the route from the instructor question preview, which will always also work in this context.

jlevine18 reviewed (commented):
removing some more changes, but otherwise good

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.html.ts:
```diff
@@ -36,109 +39,337 @@ const examplePrompts = [
           <div class="card  mb-4">
-            <div class="card-header bg-primary text-white d-flex">Generate Question using AI</div>
+            <div class="card-header bg-primary text-white d-flex justify-content-between">Generate Question using AI
+            ${qid ? html`<div class="d-flex flex-row-reverse"><button type="button" class="btn btn-sm btn-light" data-toggle="modal" data-target="#finalizeModal">Finalize question</button></div>` : ''}</div>
```
finalize modal button, only appears in editor mode (qid != null)

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.html.ts:
```diff
@@ -36,109 +39,337 @@ const examplePrompts = [
-                  Previous Generation Jobs
-                </a>
+              ${
+                !prompts || prompts.length <= 0
```
if qid is null or there's no recorded prompts for the qid, then we're in question creator [fallback] mode (see comment below)

nwalters512 replied:
This is no longer relevant after splitting the pages.

jlevine18 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.html.ts:
```diff
@@ -36,109 +39,337 @@ const examplePrompts = [
+                          <div class="col">
+                            ${prompts
+                              .filter((x) => {
+                                return x.prompt_type !== 'auto_revision';
```
right now, there's not a publicly available way to check error detection and regeneration- should we add a separate page for it?

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.ts:
```diff
@@ -0,0 +1,78 @@
+      await queryRows(
+        sql.select_draft_generation_info_by_course_id,
+        { course_id: res.locals.course.id },
+        draftMetadataWithQidSchema,
```
Conventionally this should start with a capital letter.

jlevine18 replied:
changed in 15ec6dd

jlevine18 replied:
Changed in 15ec6dda6dc5f4677ed02a739e8498c6b7fa883a.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.ts:
```diff
@@ -0,0 +1,78 @@
+        { course_id: res.locals.course.id },
+        draftMetadataWithQidSchema,
+      )
+    ).filter((x) => x.qid !== null);
```
We should do this filtering in the database instead. That said: a non-deleted question must always have a QID, so I'd be inclined to just update the results schema to not allow `qid` to be nullable.

jlevine18 replied:
Changed in c29344763ad5b8e3a9162e506b2cac86565bfa8a.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.ts:
```diff
@@ -0,0 +1,78 @@
+        }
+      }
+
+      const drafts = await queryRows(
+        sql.select_draft_generation_info_by_course_id,
+        { course_id: res.locals.course.id },
+        draftMetadataWithQidSchema,
+      );
+
+      res.send(InstructorAIGenerateDrafts({ resLocals: res.locals, drafts }));
```
A `POST` handler should always redirect rather than sending its own response (see https://en.wikipedia.org/wiki/Post/Redirect/Get).

jlevine18 replied:
Changed in c29344763ad5b8e3a9162e506b2cac86565bfa8a.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -81,8 +94,44 @@ router.get(
+        await logPageView('instructorQuestionPreview', req, res);
+        setRendererHeader(res);
+      }
+      console.log(req);
```
Whoops, I think this was my bad:

```suggestion
```

nwalters512 reviewed (commented):
Here's a first round of comments!

jlevine18 commented:
I switched to breaking up the editor and initial generator up into other pages and to using question IDs instead of (QID, course) pairs- I think I added all of the correct course checks to ensure that you can't access other courses' questions but I could be wrong. Once this gets merged, I'll make PRs for a batch-delete TRPC/editor and revert to edit/save manual revision buttons in the editor.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -0,0 +1,114 @@
+});
+export type DraftMetadataWithQid = z.infer<typeof DraftMetadataWithQidSchema>;
+
+export function InstructorAIGenerateDraftsPage({
```
We don't typically include `Page` in these names.

nwalters512 replied:
Resolved in e9257e2.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.html.ts:
```diff
@@ -0,0 +1,267 @@
+import { compiledScriptTag, nodeModulesAssetPath } from '../../../lib/assets.js';
+import { type Question, type AiGenerationPrompt } from '../../../lib/db-types.js';
+
+export function AiGenerateEditorPage({
```
This should match the name of the page as listed in the filename. We don't typically include `Page` in these function names.

nwalters512 replied:
Resolved in e9257e2.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:
```diff
@@ -82,7 +39,11 @@ router.get(
 
-    res.send(AiGeneratePage({ resLocals: res.locals }));
+    res.send(
+      AiGeneratePage({
```
This should match the name of the page as it is in the filename.

nwalters512 replied:
Resolved in e9257e2.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.html.ts:
```diff
@@ -41,106 +42,135 @@ export function AiGeneratePage({ resLocals }: { resLocals: Record<string, any> }
+                      </p>
+                      <form
+                        name="add-question-form"
+                        hx-post="${resLocals.urlPrefix}/ai_generate_question"
```
Is using HTMX really necessary here? It seems like we're just relying on a redirection after the response, which we could do without HTMX here.

nwalters512 replied:
Answered my own question: yes, we use it to show errors, I think.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts:
```diff
@@ -0,0 +1,243 @@
+  '/',
+  asyncHandler(async (req, res) => {
+    assertCanCreateQuestion(res.locals);
+    if (res.locals.question_id) {
```
What is expected to set this value? This seems like the sort of thing that should come from `req.params` instead.

nwalters512 replied:
I changed this to use `req.params` instead in 2695872.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -0,0 +1,114 @@
+  question_id: IdSchema.nullable(),
+  updated_by: IdSchema.nullable(),
+  qid: z.string(),
+});
```
Resolved in e9257e2.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -0,0 +1,114 @@
+                            : formatDate(row.created_at, resLocals.course.display_timezone)}
+                        </td>
+                        <td>${row.uid ?? '(System)'}</td>
+                        ${row.question_id && row.id
```
Resolved in e9257e2.

nwalters512 reviewed (approved):
Let's see if this works!
